### ISSUE TITLE
Promotion / create PR feature for Bitbucket server

### ISSUE THREAD

**Issue by srisek** [akuity/kargo#2569]
We are currently evaluating Kargo and we would like Kargo to have a feature to open up Pull requests to Bitbucket server for our prod staging promotion process. We see that Kargo has a feature to support GitHub and GitLab, but not BB server. Is that some thing that can be added in the future releases , that will enable us to get going with Kargo.

**Comment by krancour** [comment]
This isn't a high priority for the maintainers at the moment, but we would fully support this. It's an issue of medium difficulty and we would welcome someone from the community to pick it up and run with it.

**Comment by chamodshehanka** [comment]
I would like to take this, and thinking to use this https://github.com/ktrysmt/go-bitbucket or BitBucket API

**Comment by krancour** [comment]
No objection to that package.

**Comment by chamodshehanka** [comment]
> krancour

Awesome. I'll start working on it

**Comment by dspereira004** [comment]
> > krancour
> 
> Awesome. I'll start working on it

Hey @chamodshehanka, any news regarding the development of the Bitbucket PR functionality? Do you still plan to do it?

**Comment by chamodshehanka** [comment]
> dspereira004

yes. Can I open a draft PR?

**Comment by dspereira004** [comment]
> > dspereira004
> 
> yes. Can I open a draft PR?

That would be awesome! But, for more context: we are also evaluating Kargo for our company and our initial impression is good. However, we really need Bitbucket PR support, which is why I was asking you if you'll add support for it.

**Comment by chamodshehanka** [comment]
> > > dspereira004
> > 
> > 
> > yes. Can I open a draft PR?
> 
> That would be awesome! But, for more context: we are also evaluating Kargo for our company and our initial impression is good. However, we really need Bitbucket PR support, which is why I was asking you if you'll add support for it.

that's great! we could work together then

**Comment by krancour** [comment]
Calling attention to #3128 which is an in-progress PR to add Azure DevOps support. It's a good example of how support for a new Git provider is added.

**Comment by valkiriaaquatica** [comment]
Hi  @krancour , after reading the issue I get for Azure Devops and Bitbucket is community made right? In my case is  Gitea integration, would it follow the same proccedure?

Thanks in advanced!

**Comment by krancour** [comment]
Yes. Support for any/all Git hosting providers all should implement the same interface.

**Comment by krancour** [comment]
Not stale. In progress in #3209.

**Comment by jo8s** [comment]
I would like to reopen this issue as it looks like it was supposed to add support for Bitbucket Server, which is what I'm looking for as well. 

[#3209](https://github.com/akuity/kargo/pull/3209) seems to add support for Bitbucket Cloud only. Is this correct?

**Comment by krancour** [comment]
> https://github.com/akuity/kargo/pull/3209 seems to add support for Bitbucket Cloud only. Is this correct?

This appears to be the case.

The author of #3029 either overlooked the operative word "server," did not understand it to be a distinct thing from Bitbucket's SaaS, or, in light of the fact that we didn't have Bitbucket support _of any kind_ at the time, may have innocently misunderstood the nature of the issue to be, simply, "add Bitbucket support." 

There are comments in the code explicitly saying this doesn't work for Datacenter and I know I saw that and was ok with it at the time I signed off on #3029. In retrospect, I made the mistake of evaluating the PR in a vacuum, weighing it only on its own merits, and _also_ failing to notice the operative word "server" in this issue. 🤦‍♂️ 

Let's start with a fresh issue.

**Comment by krancour** [comment]
Opened https://github.com/akuity/kargo/issues/6020
